src/core/mac/data_poll_handler.cpp

```diff
@@ -313,6 +320,126 @@ void DataPollHandler::ProcessPendingPolls(void)
     }
 }
 
+#if OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE
+otError DataPollHandler::HandleCslFrameRequest(Mac::TxFrame &aFrame)
+{
+    otError error = OT_ERROR_NONE;
+    Child & child = *mCslTxChild;
+
+    VerifyOrExit(child.IsCslSynchronized(), error = OT_ERROR_ABORT);
+
+    OT_ASSERT(mIndirectTxChild == NULL);
+
+    // If this message is fragmented, the SSED may use data poll to request those fragments.
+    mIndirectTxChild = mCslTxChild;
+    SuccessOrExit(error = HandleFrameRequest(aFrame));
+
+    aFrame.SetChannel(child.GetCslChannel());
+    aFrame.SetTxPhase(child.GetCslPhase());
+    aFrame.SetTxPeriod(child.GetCslPeriod());
+
+exit:
+    // No matter this attempt succeeds or fails, the sender have to wait for the next period.
+    mCslTxChild = NULL;
+    return error;
+}
+
+uint32_t DataPollHandler::GetNextCslTransmitRequestDelay(const Child &aChild, uint64_t aRadioNow)
+{
+    uint32_t delay;
+    uint16_t phase = (aRadioNow / OT_US_PER_TEN_SYMBOLS) % aChild.GetCslPeriod();
+    int16_t  diff  = static_cast<int16_t>(aChild.GetCslPhase() - phase);
+
+    if (diff > kCslMaxFrameRequestAhead)
+    {
+        delay = static_cast<uint16_t>(aChild.GetCslPhase() - phase - kCslMaxFrameRequestAhead) * OT_US_PER_TEN_SYMBOLS;
+    }
+    else if (diff > kCslMinFrameRequestAhead)
```

Comment:
Do you mean make the delay so that it's in next CSL period cycle? But I'm thinking what if the MAC is still in transmitting state in next CSL period cycle. It seems that this is unavoidable. Thoughts?

Reply:
If there are two SSEDs, this way it can yield to the second SSED which might become closer to its sample window than this one, thus better scheduling.

Schedule this frame to next round of CSL period will not be a problem. It is scheduled `kCslMaxFrameRequestAhead`(8ms) before sample window. When the timer fires, if there are no ongoing transmission, it will hold the queue for 8ms. If there are ongoing transmission, then 8ms is big enough for the biggest 15.4 frame transaction to be finished. This indirect transmission will hold the queue after the direction transmission(less than 8ms) and will not encounter the problem.

Reply:
Ok, make sense.